[announcing babelfish] blogpost

abeaumont opened the pull request:
Initial blogpost draft to get feedback about contents and scope.

We may want to include some hands-on example of a babelfish tool, or just leave as it is to not make it too lengthy.

smola commented on content/upcoming/announcing_babelfish.md:
```diff
@@ -0,0 +1,82 @@
+---
+author: abeaumont
```
Please, create a PR adding yourself as blog author: https://github.com/src-d/blog/blob/master/config.yaml#L42

alcortesm commented on content/upcoming/announcing_babelfish.md:
```diff
@@ -0,0 +1,82 @@
+As shown in the figure, Babelfish is a client/server application. A client is a
+language analysis tool that wants to perform some analysis on a source code. It
+connects to the server handle source code parsing by using the needed language
+driver containers.
```
This paragraph needs rewording, something like this:
```
As shown in the figure,
Babelfish is a client/server system.
Clients are source code analysis tools that rely on the server for the parsing of the actual source code written in different programming languages.

The server itself uses language drivers to perform the parsing,
which are divided into two parts:
```

smola replied:
Maybe here `s/language driver containers/language drivers/`?
Then explain that drivers are containers where it it relevant.

smola commented on content/upcoming/announcing_babelfish.md:
```diff
@@ -0,0 +1,82 @@
+[Roadmap](https://blog.sourced.tech/post/our-roadmap/) states the first step in this
+direction as: *Build representations of source code, developers and projects.*
+
+Today we are announcing Babelfish, the project we are developing to build this
```
You could link first mention to Babelfish to https://doc.bblf.sh

smola commented on content/upcoming/announcing_babelfish.md:
```diff
@@ -0,0 +1,82 @@
+
+As shown in the figure, Babelfish is a client/server application. A client is a
+language analysis tool that wants to perform some analysis on a source code. It
+connects to the server handle source code parsing by using the needed language
```
`s/It connects to the server handle source code/It connects to the server to handle source code`?
`s/It connects to the server handle source code/It connects to the server to request parsing of a source code file`?

smola commented on content/upcoming/announcing_babelfish.md:
```diff
@@ -0,0 +1,82 @@
+  directly from target language's compiler tools or libraries.
+- An AST normalizer written in Go, that gives the UAST as a result.
+
+You can look at the [documentation](https://doc.bblf.sh/architecture.html) for
```
It would be good to add a section or paragraph about the fact that Babelfish server uses containers to be able to run drivers using disparate ecosystems with minimal overhead for the user. Make sure you include the words `libcontainer` and `Docker` in the blog post ;)

smola commented on content/upcoming/announcing_babelfish.md:
```diff
@@ -0,0 +1,82 @@
+launch the corresponding language driver and return a response with the parsed
+file.
+
+We have a dozen language drivers in various stages of development, with a quite
```
You can link `a dozen language drivers` to `https://doc.bblf.sh/languages.html`

abeaumont commented on content/upcoming/announcing_babelfish.md:
```diff
@@ -0,0 +1,82 @@
+categories: ["technical"]
+---
+
+At source{d} we believe there's a possibility for programs to write their own
```
I took that directly from [Our Roadmap blogpost](https://blog.sourced.tech/post/our-roadmap/): _At source{d} we believe that neural networks have opened the possibility for programs to write their own code and for new forms of automatic programming to emerge._
It may be better to leave the whole phrase to have a more complete context?
`At source{d} we believe there's a possibility for programs to write their own code and for new forms of automatic programming to emerge.`

eiso replied:
I see the ambiguity that @serabe mentions, "to write code" will make it clearer and broader. Thanks for catching my mistake.

abeaumont replied:
Oh, I misunderstood @Serabe. Will change it back, thanks

abeaumont commented:
Updated with received feedback.

juanjux commented on content/upcoming/announcing_babelfish.md:
```diff
@@ -0,0 +1,86 @@
+Abstract Syntax Tree [(UAST)](https://doc.bblf.sh/uast/specification.html)
+
+Well, that's our objective, we are not there just yet, but we have designed an
+architecture to achieve exactly that, and we're currently working it.
```
Maybe a little more emphasis on working? Like "currently working intensely/enthusiastically on it".

juanjux commented on content/upcoming/announcing_babelfish.md:
```diff
@@ -0,0 +1,86 @@
+written in different programming languages.
+
+The server itself uses language drivers to perform the parsing, which are
+divided in two parts, in order to minimise the work of developing a new driver:
```
Mmmm maybe it's me on monday morning but I'm not sure the "which are" have the drivers as subject or the parsing (in that case it would be "which is"). The second option sounds better to me, I think.

Edit: forget it, from the context below it's clearly referring to the drivers.

juanjux commented on content/upcoming/announcing_babelfish.md:
```diff
@@ -0,0 +1,86 @@
+divided in two parts, in order to minimise the work of developing a new driver:
+
+- A language code parser, which builds a native AST. This parser can be built
+  directly from target language's compiler tools or libraries.
```
Minor: from *the* target language

juanjux commented on content/upcoming/announcing_babelfish.md:
```diff
@@ -0,0 +1,86 @@
+frees the user from having to handle dozens of different language ecosystems,
+since the server just handles driver using Docker images.
+
+You can look at the [documentation](https://doc.bblf.sh/architecture.html) for
```
Super minor: "You can look at the documentation..." sounds a little more RTFMish and less inviting than "Take a look at the documentation..." but this is super subjective.

juanjux commented:
I like it in its current form. It's short but concise, informative and to the point. Good work!

smeruelo commented on content/upcoming/announcing_babelfish.md:
```diff
@@ -0,0 +1,86 @@
+direction as: *Build representations of source code, developers and projects.*
+
+Today we are announcing [Babelfish](https://doc.bblf.sh/), the project we are
+developing to build this representations of source code.
```
s/this/these

smeruelo commented on content/upcoming/announcing_babelfish.md:
```diff
@@ -0,0 +1,86 @@
+
+The server uses containers to run these drivers through libcontainer. This
+frees the user from having to handle dozens of different language ecosystems,
+since the server just handles driver using Docker images.
```
s/driver/drivers

smeruelo commented on content/upcoming/announcing_babelfish.md:
```diff
@@ -0,0 +1,86 @@
+[python driver](https://github.com/bblfsh/python-driver) and a usable
+[java driver](https://github.com/bblfsh/java-driver).
+
+We have a set of [tools](https://github.com/bblfsh/tools) that showcase how
```
s/showcase/showcases

smeruelo commented on content/upcoming/announcing_babelfish.md:
```diff
@@ -0,0 +1,86 @@
+babelfish works and how it can be used to build your own code analysis tools on
+top of it.
+
+## How can I contribute
```
Missing question mark!

smeruelo commented on content/upcoming/announcing_babelfish.md:
```diff
@@ -0,0 +1,86 @@
+have.
+
+If you find it useful you're more than welcome to contribute code any of its
+components, or even write a driver to your favourite language! Drop by the slack
```
s/to/for

smeruelo commented on content/upcoming/announcing_babelfish.md:
```diff
@@ -0,0 +1,86 @@
+## How can I contribute
+
+Babelfish is being openly developed. The code is available at
+[bblfsh project Github](https://github.com/bblfsh/), discussions are held at
```
"and discussions.." sounds better

smeruelo commented on content/upcoming/announcing_babelfish.md:
```diff
@@ -0,0 +1,86 @@
+(#babelfish channel).
+
+If you're interested in the project, you can look further at the
+[documentation](https://doc.bblf.sh/), give it a try, report back any issue you
```
Same here, "and report back.."

smeruelo commented on content/upcoming/announcing_babelfish.md:
```diff
@@ -0,0 +1,86 @@
+[documentation](https://doc.bblf.sh/), give it a try, report back any issue you
+have.
+
+If you find it useful you're more than welcome to contribute code any of its
```
Use a comma in conditional sentences (as you did in the previous one).
Also, the second part sounds weird. Maybe "contribute to any of its components"?

margaridag commented:
@abeaumont great post!!

My only comment would be to:
- Add at the bottom after Curry-on that Santiago is also speaking at Docker's HQ in SF (link: https://www.meetup.com/es-ES/Docker-meetups/events/240565310/?eventId=240565310&chapter_analytics_code=UA-48368587-1)

vmarkovtsev commented:
https://github.com/abeaumont/blog/pull/1

vmarkovtsev commented:
There are other two suggestions which will improve tbe post a lot. The first is the story of @smola and @juanjux about how hard it is to design and classify a UAST. Of course, an image of UAST node types is highly beneficial. The second is a few lines of code which do something useful, e.g. calculate CC using the  tools. Don't be afraid to go into deep tech details, our real users are qualified enough. From my experience, giving links docs and asking to play without any sexy in line examples is not effective.

abeaumont commented:
@margaridag sure, I will add it.
@vmarkovtsev thanks, some comments:
- It'd be better to have your change proposals in this PR, to prevent the discussion from spreading. I'll incorporate the changes I agree with and we can continue the discussion here.
- I don't know what story of @smola and @juanjux you are refering to, I can check it if you point me to it. Note that I've left any complexity discussion out on purpose, since I don't think we really know how complex it is.
- I didn't want to extend the blogpost too much, that's why I linked to the documentation where you have these lines of code. Interested people can go see them while less tech-savvy ones can still see the big picture. If people prefer to have more code samples I can rework the blogpost with this in mind. I'd personally prefer to have it published as soon as possible instead (this week, before Curry On).

vmarkovtsev commented:
@abeaumont Pushing PRs to PRs is a common way of polishing stuff in src-d/blog. Me, Marge and Eiso use it all the time, because it is easier to fix everything yourself than fuck the reviewee (besides, it is much faster, too).

vmarkovtsev commented:
@abeaumont Please don't be defensive: you've got only one attempt to announce Babelfish, that's why I am being an asshole. Of course you can ignore my experience, but it was gained with blood an sweat, and I was exactly like you several years ago.

abeaumont commented:
I see the point of the PR, but I'd like to discuss many of the points (i. e. some grammar fixes are valid, while others are unneeded and some are incorrect, I don't think we should include BIPs since we don't even have a landing page or the early python client which doesn't use the library we are designing -and it's not even merged yet-, etc.)

I'm just giving some arguments, not being defensive. That's why I want to stay with the discussion in this PR, so that everybody can share its opinion and build a better post together.

vmarkovtsev commented:
Regarding BIPs: the linked README is clear about them and how they work, and I see that you really use them. And honestly, it is the only proof of the words "open development", otherwise it is blah-blah-blah I read in every corporate post.

Python client is merged today, and your post is not published earlier then on Monday. Let me explain why. We are likely to not reach the consensus today before 16:00, and we never publish anything on Friday. So we've got the extra weekend and 2 working days - a royal gift which can be spent on adding the flesh to the post.

vmarkovtsev commented:
Regarding the links to documentation. Again, this is my bloody experience. You can consider the links to documentation never visited by 90% of your audience. If you do not catch their eye on something sexy inside the post, you lose them.

juanjux commented:
It's true that a little snippet can do wonders about firing interest among us programmers, it's one of the things that javascript frameworks do usually right. A snippet using Vadim's Python client even if it's not 100% kosher (it can't be since the libuast is not finished, which could be added as a disclaimer) could be nice.

vmarkovtsev commented:
@abeaumont Feel free to move any additions or edits from my PR without merging it.

abeaumont commented:
Rebased due to `config.yaml`.
I've included feedback from https://github.com/abeaumont/blog/pull/1 and added @margaridag's link. This is now pending on merging https://github.com/bblfsh/documentation/pull/44
I'll add a code sample if time permits (maybe we're not in such a rush as I thought since nobody has said anything so far).
About including client-python, it's my understanding that we prefer not to include it, please @mcuadros correct me if I'm wrong.

abeaumont commented:
I've added an example of how to get the server and tools up and running.

smola commented on content/post/announcing_babelfish.md:
```diff
@@ -0,0 +1,146 @@
+## How can I contribute?
+
+Babelfish's development is open
+and is based on [BIPs](https://doc.bblf.sh/proposals/README.md).
```
Please, fix the link. The correct one is https://doc.bblf.sh/proposals/

abeaumont replied:
Fixed, thanks

abeaumont commented:
@mcuadros @eiso Publish this blogpost when you consider best. Note that if published later than Monday it won't make sense to include @smola's talk at CurryOn (or should be reworded at least). If published later than Wednesday, the same applies to the Docker meetup talk.

Also note that I'll be at CurryOn on Monday and Tuesday.

vmarkovtsev commented:
@smola I see no approval, do you approve?

vmarkovtsev commented:
@abeaumont Can you please move the post to `upcoming` so that we can merge this and beta-test in #general. Upcomings are not visible on the main page, but accessible through the direct link.

abeaumont commented:
@vmarkovtsev done

vmarkovtsev commented:
(Santiago blessed me in the kitchen)

vmarkovtsev commented:
https://blog.sourced.tech/upcoming/announcing_babelfish/
